xff: add support for configuring a list of trusted CIDRs

jamesog opened the pull request:
Commit Message: xff: add support for configuring a list of trusted CIDRs

The original client IP address can be determined from the x-forwarded-for header either by a fixed number of trusted hops, or by evaluating the client IP address against a list of trusted addresses.

This adds support for configuring a list of CIDRs in the xff original IP detection extension. The remote IP address is evaluated against these, and optionally recurses through XFF to find the last non-trusted address.

Additional Description:
This feature is generally used by people with a CDN in front of their edge proxy to ensure that XFF is only parsed when the remote connection comes from a CDN server.

The behaviour of the new functionality should be the same as Nginx's `realip` module.

Disclaimer: This is my first time writing C++ so I'm not certain my changes are completely idiomatic, but I've tried to stick with existing style in the codebase. Feedback very welcome!

Risk Level: Medium
Testing: Unit tests, manual tests
Docs Changes: Updates to HTTP Connection Manager header manipulation docs, and proto docs.
Release Notes: Added to changelogs/current.yaml
Platform Specific Features: None
Fixes #21639 
Relates to #31296

jamesog commented on source/extensions/http/original_ip_detection/xff/xff.cc:
```diff
@@ -10,13 +11,43 @@ namespace Xff {
-    : xff_num_trusted_hops_(xff_num_trusted_hops) {}
+    : xff_num_trusted_hops_(xff_num_trusted_hops), recurse_(false) {}
+
+XffIPDetection::XffIPDetection(
```
This constructor (rightly) causes a coverage failure. I'm not actually sure if it's needed.

Its sibling (directly above) is only used by some tests but not directly via HCM, so perhaps it's not needed?

jamesog replied:
I've removed this constructor.

markdroth commented on api/envoy/extensions/http/original_ip_detection/xff/v3/xff.proto:
```diff
@@ -17,10 +20,42 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
-  // is not specified. See the documentation for
-  // :ref:`config_http_conn_man_headers_x-forwarded-for` for more information.
-  uint32 xff_num_trusted_hops = 1;
+  oneof xff_trusted_config {
```
It's not okay to move a pre-existing field into a oneof, since that's a breaking API change for golang.

Also, the [xDS API style guide](https://github.com/envoyproxy/envoy/blob/main/api/STYLE.md) says not to use oneof anyway.  Instead, simply add the new field next to the existing field, and document the fact that if both are set, the new field takes precedence.

jamesog replied:
Thanks, Mark. I did miss that API style document.

I'll work on updating this.

jamesog replied:
Updated.

markdroth commented on api/envoy/extensions/http/original_ip_detection/xff/v3/xff.proto:
```diff
@@ -17,10 +20,42 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
+  // If ``true``, recurse through the :ref:`config_http_conn_man_headers_x-forwarded-for`
+  // HTTP header to find the first (from the right) non-trusted IP address; otherwise
+  // the last address in ``x-forwarded-for`` is used.
+  bool recurse = 2;
```
Will all data planes always want false to be the default?  If not, consider using `google.protobuf.BoolValue` here, so that the control plane can differentiate between "explicitly set to false" and "unset, use default value".

jamesog replied:
Updated.

markdroth commented:
/lgtm api

alyssawilk commented on api/envoy/extensions/http/original_ip_detection/xff/v3/xff.proto:
```diff
@@ -22,5 +26,37 @@ message XffConfig {
   // is not specified. See the documentation for
   // :ref:`config_http_conn_man_headers_x-forwarded-for` for more information.
+  //
+  // If ``xff_trusted_cidrs`` is configured, this field is not used.
```
can we disallow it being set and reject config if it is?

jamesog replied:
I originally had this change as using a `oneof` encapsulating the original `xff_num_trusted_hops` and the new `xff_trusted_cidrs` fields, but Mark told me that would be a breaking API change and to do it this way instead. Do you know if there's another way to handle rejecting it without it being a breaking change?

markdroth replied:
I don't think we want to disallow if both the old and new fields are set, since that makes it hard for control planes to start using this new feature.

If we reject when both fields are set, then a control plane must know whether any given client supports the new field.  If it sets only the new field, old clients will not work properly, and if it sets both, new clients will break.  So the only way to transition from the old field to the new field is to either (a) wait for all clients to be upgraded before starting to use the new field or (b) have some way for the control plane to know which client supports the new field and which clients support the old field so that it can send the right config to each one -- and doing that in a way that does not break cacheability requires some dances with dynamic parameter support, which means that the clients have to be configured to tell the control plane whether or not they support the new field.

In contrast, if we allow both fields to be set and specify the precedence order, then this becomes much easier.  Control planes can simply set both fields.  Old clients will use the old field, and new clients will prefer the new field.

I think the latter approach is much cleaner and should be preferred any time we add this kind of new field in xDS.

jamesog replied:
How do you feel about the change I made since Alyssa's feedback? d976354a8906b186d4b01f2ab995594b2cca5933

She [suggested](https://github.com/envoyproxy/envoy/pull/31831#discussion_r1465551228) it should either log or reject; I chose to reject by throwing an error with a message that clearly informs that only one field can be set.

markdroth replied:
What I'm saying is that I don't think that's the behavior we want.  I think the way you had this originally was the right thing to do.

jamesog replied:
In that case I'm getting conflicting advice from two different maintainers :-)

@alyssawilk perhaps you can help decide which way I should take this based on Mark's feedback?

alyssawilk commented on api/envoy/extensions/http/original_ip_detection/xff/v3/xff.proto:
```diff
@@ -22,5 +26,37 @@ message XffConfig {
+  //
+  // If ``recurse`` is set the :ref:`config_http_conn_man_headers_x-forwarded-for`
+  // header is also evaluated against the trusted CIDR list, and the last non-trusted
+  // address is used as the original client address.
```
ooc when would we not want to recurse?

jamesog replied:
It's a good question! I added some commentary to the docs in 1f765841d3.

alyssawilk commented on api/envoy/extensions/http/original_ip_detection/xff/v3/xff.proto:
```diff
@@ -22,5 +26,37 @@ message XffConfig {
+  XffTrustedCidrs xff_trusted_cidrs = 2;
+}
+
+message XffTrustedCidrs {
```
I think any updates in this area of API should also be spelled out in docs/root/configuration/http/http_conn_man/headers.rst

that's probably a better place to answer my recurse question too btw :-)

jamesog replied:
I added a small note to those docs mentioning to use the extension if you need to use trusted CIDRs, but I wasn't sure how much "non-core" information should go into those docs when it otherwise talks about configuration that's directly in HCM.

Do you think I should add more than the note, or expand that note?

mattklein123 replied:
All of this code/config is so confusing, and such a security landmine, that it's quite hard to wrap my head around what this all does and why I would use it.

Can you add significant fleshed out examples either here or better in https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_conn_man/headers#config-http-conn-man-headers-x-forwarded-for where there are other examples? This would allow us to see various worked examples of what the config does and why I would want to use it, and then we can review the implementation? Thank you.

/wait

jamesog replied:
Yep, absolutely! I've been working on that recently and trying to do some more local testing to make sure I've really got everything right based on Alyssa's feedback.

The TL;DR is to make it work in a similar way to Nginx's [realip module](https://nginx.org/en/docs/http/ngx_http_realip_module.html) as per #21639.

Just to confirm something:

> All of this code/config is so confusing, and such a security landmine, that it's quite hard to wrap my head around what this all does and why I would use it.

Are you saying that my code is confusing, or the existing xff handler, or both? If it's mine I'd love to know how I can make it less confusing. I also posted a (slightly long-winded) question in Slack the other day, if either of you have a chance to take a look at that.

mattklein123 replied:
No it isn't your code, it's this entire part of the codebase. Any time anyone touches it we go around in circles because it is so confusing and the security implications are so profound. Clear examples and a lot of documentation helps a lot. Thank you!

alyssawilk commented on api/envoy/extensions/http/original_ip_detection/xff/v3/xff.proto:
```diff
@@ -22,5 +26,37 @@ message XffConfig {
+  // `CDN <https://en.wikipedia.org/wiki/Content_delivery_network>`_.
+  //
+  // If this field is configured, ``xff_num_trusted_hops`` is not used.
+  XffTrustedCidrs xff_trusted_cidrs = 2;
```
may be worth a (presumably fairly obvious) comment that these checks are n*m where n is the number of trusted ranges and m is the number of xff hops in the incoming header.

jamesog replied:
Good call, added in eb2f7ff10f.

alyssawilk commented on source/common/http/utility.cc:
```diff
@@ -781,6 +782,61 @@ void Utility::sendLocalReply(const bool& is_reset, const EncodeFunctions& encode
+    }
+    return {addr, true};
+  }
+  return {nullptr, false};
```
if every address in the XFF is trusted don't we want to return the last address in the list?

jamesog replied:
This one made me think! Originally I was assuming this might be some kind of undefined behaviour and so it was safer to return null, but after talking about this with a colleague, and looking over what other proxy servers do, I implemented your suggestion in a016a6cf94. Thanks for questioning my logic :-)

alyssawilk reviewed (commented):
So sorry for the review delay.  Thanks for the well documented and tested first pass - my thoughts below.
/wait

alyssawilk commented on source/common/http/utility.cc:
```diff
@@ -781,6 +782,65 @@ void Utility::sendLocalReply(const bool& is_reset, const EncodeFunctions& encode
+      // or XFF is malformed.
+      return {nullptr, false};
+    }
+    return {addr, true};
```
I don't think we want to just return true here or below.  I believe the final boolean here is "do we trust the whole chain such that we can say this is an internal request".   I don't _think_ we want to equate "we have a chain of trusted cidr addresses" with "we have a trusted chain of internal addresses".  I think we should only allow treating the chain as internal iff the direct connected host is internal and every hop is not only cidr-trusted but internal.

jamesog replied:
Oof, you're right, this is definitely the wrong thing to do here. And now the more I read on what "internal" means in Envoy the more confused I'm getting 😅. The return in `getLastAddressFromXFF` is also pretty confusing.

I've read through some history in #2559 and #2590 to try and figure out what the correct answer is here. What should "internal" mean when you're saying "I trust these CIDRs"? Internal doesn't always have to mean RFC1918 or ULA, for example.

To give a concrete example for my use-case for using this feature, I want to add my CDN's ranges, and our own internal proxy IPs (RFC1918), but _not_ the whole internal network; anything else internal should be untrusted.

If Envoy's definition of "internal" is simply "RFC1918 or ULA", perhaps while looping through `trusted_cidrs`:
- Set a boolean marking if any CIDR is not internal;
- On this line I can exchange `true` with `isInternalAddress(addr)`;
- On the final line use the bool from point 1

WDYT? I'm still not convinced I've got this entirely right.

alyssawilk replied:
for "internal" I usually read "sent by someone in my company", like I trust them to hit up /quitquitquit
for "trusted" here that's "trusted to set XFF header"
so unless you trust your CDN to verify "this is actually someone from my company sending traffic through the CDN" I'd say "not internal" :-)

jamesog replied:
How do you think we should mark that with a list of CIDRs? I could split it into "internal_cidrs" and "external_cidrs", which feels a bit clunky but would do the job.

jamesog replied:
I've dug through how the internal address config works, and how the `detect()` method is called on the xff extension, and my understanding is:

- The bool here is `allow_trusted_address_checks`, not "do we trust this address": https://github.com/envoyproxy/envoy/blob/ac42a62d6b8a2599c9a6101b6bb0bd11220c63e6/envoy/http/original_ip_detection.h#L36-L38
- The `allow_trusted_address_checks` bool determines whether to compare the address against `internal_address_config` in HCM: https://github.com/envoyproxy/envoy/blob/ac42a62d6b8a2599c9a6101b6bb0bd11220c63e6/source/common/http/conn_manager_utility.cc#L217-L228

So given that, I think returning `true` is the correct thing to do, and then HCM will compare the address returned against its internal address config to really determine if it's internal. Given that, it fits my use-case described above perfectly and there's no need to split the list or do extra checks here.

Does my reasoning seem sound?

alyssawilk commented on source/common/http/utility.cc:
```diff
@@ -781,6 +782,65 @@ void Utility::sendLocalReply(const bool& is_reset, const EncodeFunctions& encode
+    }
+    last_valid_addr = addr;
+
+    bool trusted = false;
```
remoteAddressIsTrustedProxy?

alyssawilk reviewed (commented):
Looking good!  I have a few clarifying points and once we settle this round I should be good to take a look at tests as well
/wait

jamesog commented on source/extensions/http/original_ip_detection/xff/xff.cc:
```diff
@@ -10,13 +11,33 @@ namespace Xff {
 XffIPDetection::XffIPDetection(
     const envoy::extensions::http::original_ip_detection::xff::v3::XffConfig& config)
-    : xff_num_trusted_hops_(config.xff_num_trusted_hops()) {}
+    : xff_num_trusted_hops_(!config.has_xff_trusted_cidrs() ? config.xff_num_trusted_hops() : 0),
```
That makes sense, thanks!

jamesog commented:
Apologies for the delay in pushing updates. I got caught in the weeds a bit trying to understand some of the internals.

I think I've addressed all of Alyssa's feedback, and I added more documentation and examples to headers.rst.

The one change I'm not sure about is a587a23d55ca8d042250c87b33d1d7db3af4f138 - I've tried to understand what "internal" means to Envoy and I still can't say I'm 100% sure. PR #2232 added some comments about it needing to be a single IP in XFF, so I've made a similar change here but I don't feel quite right about it.

mattklein123 reviewed (changes requested):
Thanks, makes sense to me at a high level. A couple of questions from a first pass. Thank you!

/wait

jamesog commented on source/common/http/conn_manager_utility.cc:
```diff
@@ -172,6 +172,12 @@ ConnectionManagerUtility::MutateRequestHeadersResult ConnectionManagerUtility::m
       if (result.reject_options.has_value()) {
         return {nullptr, result.reject_options};
       }
+      if (Network::Utility::isLoopbackAddress(
+              *connection.connectionInfoProvider().remoteAddress())) {
+        Utility::appendXff(request_headers, config.localAddress());
+      } else {
+        Utility::appendXff(request_headers, *connection.connectionInfoProvider().remoteAddress());
+      }
```
I took a closer look and found that appending to XFF only happens when you use the HCM `use_remote_address` setting - there was no way to append to XFF when using the original IP detection extension.

mattklein123 replied:
So then what does the config you added do?

jamesog replied:
🤦  I should be checking for `result.append_xff` here. Thanks!

mattklein123 replied:
Please add some kind of test that covers this that fails before fixing.

jamesog replied:
Will do. I tried looking at adding a test for this before but was struggling to wrap my head around the integration tests.

jamesog replied:
Fixed this and added tests.

jamesog commented on api/envoy/extensions/http/original_ip_detection/xff/v3/xff.proto:
```diff
@@ -22,5 +26,49 @@ message XffConfig {
+  // Whether or not to recurse depends on whether you trust the downstream proxy to have
+  // validated all addresses in the list.
+  // If you do not fully trust the downstream proxy you may not wish to recurse.
+  google.protobuf.BoolValue recurse = 2;
```
Updated.

jamesog commented on api/envoy/extensions/http/original_ip_detection/xff/v3/xff.proto:
```diff
@@ -22,5 +26,49 @@ message XffConfig {
+  XffTrustedCidrs xff_trusted_cidrs = 2;
+
+  // Whether to append the detected trusted client IP address to ``x-forwarded-for``.
+  google.protobuf.BoolValue append_xff = 3;
```
Updated.

jamesog commented on docs/root/configuration/http/http_conn_man/headers.rst:
```diff
@@ -265,6 +265,20 @@ additional addresses from XFF:
+
+ If the trusted client address should be determined from a list of known CIDRs, use the
+ :ref:`xff <envoy_v3_api_msg_extensions.http.original_ip_detection.xff.v3.XffConfig>` original IP
+ detection extension instead.
```
Updated.

mattklein123 commented on api/envoy/extensions/http/original_ip_detection/xff/v3/xff.proto:
```diff
@@ -22,5 +26,46 @@ message XffConfig {
+  // If this field is configured, ``xff_num_trusted_hops`` is not used.
+  //
+  // Note: Use of this option is an ``n*m`` operation on every request, where ``n`` is
+  // the number of trusted CIDRs and ``m`` is the number of entries in the XFF header.
```
Ping on this comment?

jamesog replied:
Thanks, I did miss this. I've moved the comment.

mattklein123 commented on api/envoy/extensions/http/original_ip_detection/xff/v3/xff.proto:
```diff
@@ -22,5 +24,49 @@ message XffConfig {
+  // the number of trusted CIDRs and ``m`` is the number of entries in the XFF header.
+  XffTrustedCidrs xff_trusted_cidrs = 2;
+
+  // Whether to append the detected trusted client IP address to ``x-forwarded-for``.
+  bool append_xff = 3;
```
Sorry for being dense and asking this now, but don't we already append to XFF somewhere in the request pipeline? Does this mean that XFF will get double appended in some cases? If this is somehow different can you spell out how this is different and when this would be used?

jamesog replied:
HCM will append to XFF only when `use_remote_address` is true: https://github.com/envoyproxy/envoy/blob/bc0190146f08458ac5a58af5b3545b24660fa8f8/source/common/http/conn_manager_utility.cc#L136-L143

When an original IP detection extension is used instead - which is incompatible with `use_remote_address` we never hit that code path, so there was no way to append to XFF from an original IP extension before this change.

mattklein123 replied:
I see OK makes sense, thanks.

alyssawilk replied:
maybe worth commenting in the API docs, as I had the same question.   "this only applies when use_remote_address is false, otherwise [existing config field] applies"

mattklein123 reviewed (changes requested):
Thanks some more questions.

/wait

mattklein123 commented:
Can you merge main?

/wait

jamesog commented:
> Can you merge main?

Done.

mattklein123 commented:
/retest

mattklein123 reviewed (approved):
LGTM thank you for your patience getting this done!

mattklein123 commented:
Argh sorry I think the CI issue is real since some of this code goes in the mobile build which doesn't use exceptions. Can you take a look? Hopefully it will be an easy fix.

/wait

jamesog commented:
> Argh sorry I think the CI issue is real since some of this code goes in the mobile build which doesn't use exceptions. Can you take a look? Hopefully it will be an easy fix.

Ah, I meant to follow up on this one, I noticed that failure too. I could use some advice here actually. In an earlier change I had the constructor for the xff extension forcibly setting `xff_num_trusted_hops` to 0 when `xff_trusted_cidrs` was set, but Alyssa recommended logging an exception instead.

What's the best way to fix this to work for mobile, do you think? Should I set it back to overwriting the value and logging it, or is there another way to make it fail that works for Mobile?

mattklein123 commented:
> What's the best way to fix this to work for mobile, do you think? Should I set it back to overwriting the value and logging it, or is there another way to make it fail that works for Mobile?

Probably the best way would be to somehow make this code return an error and then somehow fail with an exception outside in the server (non mobile code) but we probably will need to ask @alyssawilk about this tomorrow when she is around.

jamesog commented:
Digging through the codebase a bit, I think I could either:

- `ENVOY_LOG(warn, "setting both xff_num_trusted_hops and xff_trusted_cidrs is not recommended")`; or
- Revert some of d976354a8906b186d4b01f2ab995594b2cca5933 and:
  - Force set `xff_num_trusted_hops_` to 0 when `has_xff_trusted_cidrs()`
  - `ENVOY_LOG(error, "cannot set both xff_num_trusted_hops and xff_trusted_cidrs; xff_num_trusted_hops is ignored")`

jamesog commented:
Hey folks, any feedback to which approach I should take here? Thanks!

jamesog commented:
Not stale, waiting for feedback! But I'm also trying to find other examples in the codebase that do similar things, but there's a lot of code to examine...

robd003 commented:
Can we keep this open and get it reviewed? @alyssawilk @mattklein123

jamesog commented:
I put it back into draft because I need to make some more changes but I haven't had the time. I'm hoping to get back to it soon.

robd003 commented:
Please keep this open, also Happy 4th of July! 🇺🇸

jamesog commented:
Sorry for the long delay here. Life kind of got in the way and it took me some time to trawl through existing code to see if there was an existing pattern for handling the case on mobile to remove the throw. All tests are passing now, hopefully this covers the last of the feedback.

mattklein123 reviewed (approved):
Thanks this LGTM. Just because this code is so scary I'm going to ask @alyssawilk to do a final pass on this. Thanks.

alyssawilk commented on source/common/http/conn_manager_utility.cc:
```diff
@@ -171,6 +171,14 @@ ConnectionManagerUtility::MutateRequestHeadersResult ConnectionManagerUtility::m
         return {nullptr, result.reject_options};
       }
+      if (result.append_xff) {
+        if (Network::Utility::isLoopbackAddress(
```
maybe worth a wrapper utility given this looks like a copy paste of preexisting code?

jamesog replied:
I've factored this code out to a wrapper function.

alyssawilk commented on source/common/http/utility.cc:
```diff
@@ -756,6 +757,65 @@ void Utility::sendLocalReply(const bool& is_reset, const EncodeFunctions& encode
+  Network::Address::InstanceConstSharedPtr last_valid_addr;
+
+  for (auto it = xff_entries.rbegin(); it != xff_entries.rend(); it++) {
+    auto addr = Network::Utility::parseInternetAddressNoThrow(std::string(*it));
```
should we allow invalid entries?  I'd be mildly inclined to fail out.

jamesog replied:
Agreed, I'll make that change.

jamesog replied:
Updated.

alyssawilk commented on source/common/http/utility.cc:
```diff
@@ -756,6 +757,65 @@ void Utility::sendLocalReply(const bool& is_reset, const EncodeFunctions& encode
+        Network::Utility::isLoopbackAddress(*addr.get())) {
+      // If the next IP after skipping all trusted IPs is internal or loopback
+      // we cannot determine the true remote IP: Either a proxy didn't add its IP,
+      // or XFF is malformed.
```
sorry probably naive here but if every hop is trusted do we not trust the whole set?
Agree if we have trusted, untrusted, trusted I wouldn't trust the final trusted.

jamesog replied:
The aim is to find the first non-trusted public IP. The condition here says if we treated all previous IPs as proxies we trust, but we then hit a private IP, we can't determine the "real" client IP so we bail out.

jeffgran-dox replied:
I think in that case the "real" IP would be the private IP, right? I think that's a valid case, where the end client that the traffic originated from was an internal IP.

jamesog replied:
The "real real" IP may well be a private IP, yes, but a private IP isn't useful to an Internet-facing proxy and, AFAIK, no implementation will consider private IPs. The exception I've seen is Apache's mod_remoteip which has a separate CIDR list for internal proxies.

If we want to consider allowing private IPs as being a real client IP I would prefer to gate that behind another setting.

alyssawilk replied:
how about we at least call that out in the API docs.  I've done my fair share of telneting into localhost for debug so at least we can hopefully avoid surprising anyone that way.

jamesog replied:
I another round of consideration. First, I was wrong on

> AFAIK, no implementation will consider private IPs

I couldn't remember where I got that from, so I went and checked a few implementations and none of them have a check like this. I suspect I got myself muddled over the general view that private IPs aren't useful (for Internet-facing proxies).

Second, on balance having this check is far too restrictive and makes assumptions about the use cases. As Alyssa alluded to, if we trust a proxy, we should trust that it validated the header.

Therefore, I've removed this block and updated the API docs to remove mention of that restriction.

alyssawilk reviewed (commented):
Hey sorry just back from vacation sorry about delay.  thoughts below
/wait

jamesog commented on api/envoy/extensions/http/original_ip_detection/xff/v3/xff.proto:
```diff
@@ -22,5 +24,51 @@ message XffConfig {
+  //
+  // Note: Use of this option is an ``N*M`` operation on every request, where ``N`` is
+  // the number of trusted CIDRs and ``M`` is the number of entries in the XFF header.
+  bool recurse = 2;
```
> could we kill this option off entirely and say the trusted proxy should sanitize the xff headers to only addresses we want to use?

I've had some discussion with my colleagues and we concur, so I'll remove this option and change the extension to always recurse when using trusted CIDRs.

This implementation was based on how Nginx does it, but your question made me do some more review and it looks like Nginx is an outlier here and basically everything else doesnt't provide that option and always does the recursive search.

Thanks for making me think more about this!

jamesog replied:
This option has been removed.

jamesog commented on source/extensions/http/original_ip_detection/xff/xff.cc:
```diff
@@ -11,7 +11,9 @@ namespace Xff {
     const envoy::extensions::http::original_ip_detection::xff::v3::XffConfig& config)
-    : xff_num_trusted_hops_(config.xff_num_trusted_hops()), append_xff_(config.append_xff()) {
+    : xff_num_trusted_hops_(config.xff_num_trusted_hops()),
+      skip_xff_append_(PROTOBUF_GET_WRAPPED_OR_DEFAULT(
```
Sorry to introduce yet another new change.

In updating the proto docs for `append_xff` I realised it should probably be named consistently with the equivalent HCM option, `skip_xff_append`, so I've changed that, but this has a somewhat unfortunate consequence: Ideally Envoy should always append XFF by default, but the xff extension currently doesn't do that and there are dozens of tests that rely on this behaviour, and I imagine Envoy users also may be depending on this behaviour.

I've therefore made this default to true for users of the existing `xff_num_trusted_hops` setting, but for the new `xff_trusted_cidrs` option I would prefer to disable this setting by default to provide proper proxy behaviour out of the box.

Let me know if you'd prefer I back this out or make the setting consistent. The current behaviour is documented in the proto so it's prominent in the docs.

alyssawilk replied:
ugh, yeah.  Generally I'm in favor of "fixing" broken or inconsistent config with a runtime guard for temporary workaround but I think in this case the risk is too high.  Checking with @mattklein123 for a second opinion but I'm inclined to agree on playing it safe over consistent config.

alyssawilk commented on source/common/http/utility.cc:
```diff
@@ -756,6 +757,65 @@ void Utility::sendLocalReply(const bool& is_reset, const EncodeFunctions& encode
+    return {nullptr, false};
+  }
+
+  absl::string_view xff_string(xff_header->value().getStringView());
```
request_headers.getForwardedForValue() and you can skip the null pointer checks above

alyssawilk commented on source/common/http/utility.cc:
```diff
@@ -756,6 +757,65 @@ void Utility::sendLocalReply(const bool& is_reset, const EncodeFunctions& encode
+  }
+  // If we reach this point all addresses in XFF were considered trusted, so return
+  // first IP in XFF (the last in the reverse-evaluated chain).
+  return {last_valid_addr, xff_entries.size() == 1};
```
do you have an e2e test for a present but empty header?

jamesog replied:
Ooh that's an interesting one. I've never seen that in the wild and I'm not totally sure it's valid, but I'll add that test.

jamesog replied:
I've added the test to cover the empty header value case.

It still presumes that we should return false for the second parameter here, based on [this comment](https://github.com/envoyproxy/envoy/blob/603bc61fbbf0ee968da5357bb258221609dc48ba/source/common/http/conn_manager_utility.cc#L214-L222), particularly

>   // 1) After remote address/XFF appending, the XFF header must contain a *single* address.

alyssawilk commented on api/envoy/extensions/http/original_ip_detection/xff/v3/xff.proto:
```diff
@@ -22,5 +26,53 @@ message XffConfig {
+  // is false, otherwise :ref:`skip_xff_append
+  // <envoy_v3_api_field_extensions.filters.network.http_connection_manager.v3.HttpConnectionManager.skip_xff_append>`
+  // applies.
+  google.protobuf.BoolValue skip_xff_append = 3;
```
mild preference for making the new fields not-implemented/hide for one release cycle with an explicit date TODO to remove it, just to give you some production time to work any kinks out before making changes is super high risk.

wbpcode replied:
I am fine to this API because this is a new feature to the Xff extension and won't effect exist behaviors.

alyssawilk commented on source/common/http/utility.h:
```diff
@@ -389,6 +390,26 @@ struct GetLastAddressFromXffInfo {
+                                 const std::vector<Network::Address::CidrRange> trusted_cidrs);
+
+/**
+ * Retrieves the last address in the x-forwarded-header after removing all trusted proxy addresses.
```
again please call out exemptions (localhost/private)

jamesog replied:
As above I've removed the restriction now.

alyssawilk commented:
hm, CI looks wedged.  can you do a main merge - maybe sorting current.yaml will fix?

alyssawilk reviewed (commented):
largely LGTM.  @wbpcode can you do a second pass?

jamesog commented on source/common/http/utility.cc:
```diff
@@ -756,6 +757,53 @@ void Utility::sendLocalReply(const bool& is_reset, const EncodeFunctions& encode
+  const auto xff_header = request_headers.getForwardedForValue();
+  static constexpr absl::string_view separator(",");
+
+  const auto xff_entries = StringUtil::splitToken(xff_header, separator, false, true);
```
Do you think it's worth having this configurable (with a default) or having it hard-coded with a comment in the API docs?

alyssawilk replied:
we've got other hard-coded limits in this file so I think it's fine to start with a high bound and extend iff anyone asks for it.  I can think of no valid use case where hundreds of XFF headers are allowed and I've seen M*N bugs like that melt down in production when someone either screws up or attacks with many-tokened headers.

jamesog replied:
Makes sense, thanks.

jamesog replied:
Added this check with a trace-level log for debugging purposes.

jamesog commented on source/common/http/utility.cc:
```diff
@@ -756,6 +757,53 @@ void Utility::sendLocalReply(const bool& is_reset, const EncodeFunctions& encode
+    }
+    last_valid_addr = addr;
+
+    bool remoteAddressIsTrustedProxy = false;
```
Fixed, thanks.

jamesog commented:
/retest

jamesog commented:
Maybe a maintainer needs to run that? :-) Seems like CI got wedged.

adisuissa commented:
/retest

wbpcode commented on api/envoy/extensions/http/original_ip_detection/xff/v3/xff.proto:
```diff
@@ -22,5 +26,49 @@ message XffConfig {
+  //
+  // .. attention::
+  //
+  //   For proper proxy behaviour it is not recommended to set this option.
+  //   For backwards compatibility, if this option is unset and
+  //   ``xff_num_trusted_hops`` is set, it defaults to true. When
+  //   ``xff_trusted_cidrs`` is set, it defaults to false.
```
Different default values only brings confusion. Let's only use a default value with `true`.

jamesog replied:
Please see the related conversation below: https://github.com/envoyproxy/envoy/pull/31831#discussion_r1706979967

@alyssawilk seemed half inclined to go with this split default, but let me know if you still think we should really keep one default.

`true` is a bad default, to be honest, but I also don't want to break any existing users of `xff_num_trusted_hops`.

wbpcode replied:
From API's perspective, explicit and predictable behavior is always better. Different default values is always not recommend for our API.

jamesog replied:
Got it. I've updated this to default to true.

wbpcode commented on source/common/http/utility.cc:
```diff
@@ -756,6 +757,62 @@ void Utility::sendLocalReply(const bool& is_reset, const EncodeFunctions& encode
   encodeLocalReply(is_reset, std::move(prepared_local_reply));
 }
 
+bool Utility::remoteAddressIsTrustedProxy(
+    const Envoy::Network::Address::InstanceConstSharedPtr& remote,
+    const std::vector<Network::Address::CidrRange> trusted_cidrs) {
+  for (const auto& cidr : trusted_cidrs) {
+    if (cidr.isInRange(*remote.get())) {
+      return true;
+    }
+  }
+  return false;
+}
```
`const Envoy::Network::Address::Instance& remote`: const reference to shared pointer here make no sense.
`absl::Span<const Network::Address::CidrRange>`: to avoid copy whole vector. (or vector reference should be used.)

jamesog replied:
Thank you, these have been updated.

wbpcode commented on source/common/http/utility.cc:
```diff
@@ -756,6 +757,62 @@ void Utility::sendLocalReply(const bool& is_reset, const EncodeFunctions& encode
+
+Utility::GetLastAddressFromXffInfo Utility::getLastNonTrustedAddressFromXFF(
+    const Http::RequestHeaderMap& request_headers,
+    const std::vector<Network::Address::CidrRange> trusted_cidrs) {
```
ditto. Please use span or vector reference.

jamesog replied:
Updated, thanks.

wbpcode commented on source/common/http/utility.cc:
```diff
@@ -756,6 +757,62 @@ void Utility::sendLocalReply(const bool& is_reset, const EncodeFunctions& encode
+  static constexpr absl::string_view separator(",");
+  static constexpr size_t MAX_ALLOWED_XFF_ENTRIES = 20;
+
+  const auto xff_entries = StringUtil::splitToken(xff_header, separator, false, true);
```
which type is used here to store the splitted tokens? Seems `auto` is not appropriate here. Maybe you should specify a type here. InlineVector<absl::string_view, 8> will be recommend.

jamesog replied:
Updated. I used `std::vector<absl::string_view>` as this seemed to be the style elsewhere and I couldn't find any reference of `InlineVector`.

wbpcode replied:
`absl::InlinedVector` provide better performance when we only has limited hops (in most cases?). And sorry for the spelling error.

jamesog replied:
Thanks for the explanation. I've updated to use InlinedVector now.

wbpcode commented on source/extensions/http/original_ip_detection/xff/xff.cc:
```diff
@@ -10,16 +11,49 @@ namespace Xff {
+    : xff_num_trusted_hops_(config.xff_num_trusted_hops()),
+      skip_xff_append_(PROTOBUF_GET_WRAPPED_OR_DEFAULT(
+          config, skip_xff_append, config.has_xff_trusted_cidrs() ? false : true)) {
+  if (config.has_xff_trusted_cidrs() && config.xff_num_trusted_hops() > 0) {
+    ENVOY_LOG(warn, "Both xff_num_trusted_hops and xff_trusted_cidrs are configured; only "
+                    "xff_trusted_cidrs will be used");
+  }
```
It should be safe to reject the configuration?

jamesog replied:
I had previously tried to do that by `throw`ing, but this causes failures for Mobile. Is there another way to reject the configuration?

wbpcode replied:
Maybe you can should use `throwExceptionOrPanic` method for now.

jamesog replied:
Is panicking acceptable for Mobile?

jamesog replied:
I've added the `throwExceptionOrPanic`. Hopefully this won't cause the Mobile tests to fail this time. Thanks for the pointer.

wbpcode reviewed (commented):
Thanks for the contribution. Some comments are added.

wbpcode commented:
/wait

jamesog commented:
/retest

wbpcode commented:
Sorry for the delay. Some times I may ignore the notification from tools. It's OK to ping me directly on slack.

wbpcode commented:
/wait

wbpcode commented on source/common/http/utility.cc:
```diff
@@ -756,6 +757,62 @@ void Utility::sendLocalReply(const bool& is_reset, const EncodeFunctions& encode
   encodeLocalReply(is_reset, std::move(prepared_local_reply));
 }
 
+bool Utility::remoteAddressIsTrustedProxy(
+    const Envoy::Network::Address::InstanceConstSharedPtr remote,
```
See, basically we only use the smart point as parameter when we wants to change the ownership. But in this case, if we ensure this remote always be valid, then `const Envoy::Network::Address::Instance&` should be used.

If we can ensure this remote always be valid, then `OptRef<const Envoy::Network::Address::Instance>` should be used.

I think there should be the first case.

wbpcode replied:
/wait

jamesog replied:
I think I understood you correctly. This is getting beyond my knowledge here so I wasn't quite sure. Please see the latest commit that now uses `const Network::Address:Instance&`.

wbpcode reviewed (commented):
Thanks for the great contribution. Only last comment, I think.

wbpcode reviewed (approved):
LGTM. Thanks.
